stages/org.osbuild.ostree.deploy

```diff
@@ -104,44 +125,79 @@ def make_fs_identifier(desc):
     raise ValueError("unknown rootfs type")
 
 
-def main(tree, inputs, options):
-    osname = options["osname"]
-    rootfs = options.get("rootfs")
-    mounts = options.get("mounts", [])
-    kopts = options.get("kernel_opts", [])
-    ref = options["ref"]
-    remote = options.get("remote")
+def ostree_commit_deploy(tree, inputs, osname, remote, ref, kopts):
+    if len(inputs) == 0:
+        if not ref:
+            raise ValueError("ref should be specified in options")
+    elif len(inputs) == 1:
+        if ref:
+            raise ValueError("Should not specify ref if input was specified")
 
-    # If provided an input then do the pull into the tree
-    if len(inputs) != 0:
+        # If we have an input then we need to pull_local() from the input
+        # first before we deploy.
         source_repo, commits = ostree.parse_input_commits(inputs["commits"])
         target_repo = f"{tree}/ostree/repo"
         for commit, data in commits.items():
-            loopref = data.get("ref", commit)
-            ostree.pull_local(source_repo, target_repo, remote, loopref)
+            ref = data.get("ref", commit)
+            ostree.pull_local(source_repo, target_repo, remote, ref)
 
     if remote:
         ref = f"{remote}:{ref}"
 
-    kargs = []
+    kargs = [f'--karg-append={v}' for v in kopts]
+    ostree.cli("admin", "deploy", ref,
+               *kargs, sysroot=tree, os=osname)
+
+
+def ostree_container_deploy(tree, inputs, osname, kopts):
+    images = containers.parse_containers_input(inputs)
+    for image in images.values():
+        with containers.container_source(image) as (image_name, image_source):
+            extra_args = []
+            imgref = f"ostree-unverified-image:{image_source}"
+
+            extra_args.append(f'--imgref={imgref}')
+            extra_args.append(f'--stateroot={osname}')
+            extra_args.append(f'--target-imgref=ostree-unverified-registry:{image_name}')
```

Comment:
I would like to avoid hardcoding `ostree-unverified-registry`.   Most especially for the `--target-imgref`.

It's there to add pain to remind people that they should be pulling containers signed via e.g. sigstore at least.

I think the higher level input should be forced to specify this spec.

Because actually today for FCOS/Silverblue/etc, the recommendation really is `ostree-remote-registry:fedora:quay.io/fedora/fedora-coreos:stable` because that will leverage the GPG signatures we already ship.

(But as soon as you create a *derived* image, this becomes worthless and one needs to turn to container-native signature tools)

Of course, we should also be signing the Fedora containers with sigstore...

Reply:
I think it's likely that the osbuild code that fetches/caches containers will need to strip the ostree transport bits...and then we get into the messy topic of parsing those.  There's some Go code in https://github.com/coreos/rpmostree-client-go/ and obviously the native code is Rust (although I need to move that code to https://docs.rs/ostree/latest/ostree/ and not https://docs.rs/ostree/latest/ostree-ext).

I can look at trying to do a Python helper for parsing.

Reply:
> I would like to avoid hardcoding `ostree-unverified-registry`. Most especially for the `--target-imgref`.

Concretely, what should we put here instead?

Reply:
I wonder if this is really an issue.  We can verify the signatures while pulling from the container store into the local archive.  After that, it's a local store controlled by osbuild.

We make the same assumptions when pulling and embedding containers.

Reply:
I think the issue is that the qcow that then gets booted is set to follow an `ostree-unverified-registry`. `--target-imgref` is used to tell it what we want to follow for updates.

I think colin is saying we want the registry we follow for updates to try to have integrity checking turned on.

Reply:
Ah, then in that case, yes, we should figure out how to do this.

> > I would like to avoid hardcoding `ostree-unverified-registry`. Most especially for the `--target-imgref`.
> 
> Concretely, what should we put here instead?

It should be part of the source options and it will need to carry over to the deploy input options.

Reply:
but we're reusing the existing skopeo input/source for this so it would need to make sense in a container context too.

anyways, I think we can probably deal with this in a followup and just hardcode the unverified registry bits for now (AFAIK we don't have containers signed in the registry).

Reply:
> (AFAIK we don't have containers signed in the registry).

```
$ cat /etc/containers/policy.json 
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports": {
        "docker": {
	    "registry.access.redhat.com": [
		{
		    "type": "signedBy",
		    "keyType": "GPGKeys",
		    "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
		}
	    ],
	    "registry.redhat.io": [
		{
		    "type": "signedBy",
		    "keyType": "GPGKeys",
		    "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
		}
	    ]
	},
        "docker-daemon": {
	    "": [
		{
		    "type": "insecureAcceptAnything"
		}
	    ]
	}
    }
}
```

Now what we really want here is something like https://github.com/containers/skopeo/issues/1829 to help streamline this.

(And see that issue for the rationale for why I introduced these signature scheme; I know it's annoying because it's not something the rest of the container stack does)

Reply:
> (AFAIK we don't currently have containers signed in the registry)

I was referring to what we have todday for FCOS since `quay.io/fedora/fedora-coreos:stable` was what was being used in the examples. Definitely have it for the RH registries.

TBH: it sounds like what we need here is skopeo pieces in OSBuild to take a policy and a gpgkey (i.e. something that can be provided by the user).

Given that does it still make sense that we wouldn't need to block this on that and do it in a followup when those pieces land?

Reply:
> I was referring to what we have today for FCOS since quay.io/fedora/fedora-coreos:stable was what was being used in the examples added in this PR. 

We do sign that container - with ostree!   See [the example image reference in this comment](https://github.com/osbuild/osbuild/pull/1402#discussion_r1357370612).

Reply:
In a nutshell the reason for this is really the same reason dpkg/rpm check gpg signatures on packages.  It has to be on by default.  Yes, signatures are hard and people sometimes just punt and set e.g. `gpgcheck=0` in repo files.  Hardcoding `ostree-unverified-registry:` has the same smell.

Ultimately TLS is a pretty strong baseline.  But still, in case you didn't click through the previous comment, see:
https://blog.lightspin.io/aws-ecr-public-vulnerability

Reply:
> > I was referring to what we have today for FCOS since quay.io/fedora/fedora-coreos:stable was what was being used in the examples added in this PR.
> 
> We do sign that container - with ostree! See [the example image reference in this comment](https://github.com/osbuild/osbuild/pull/1402#discussion_r1357370612).

👍 - yeah I knew we still had the commit signed inside, I thought you were referring to something else. and I guess `ostree-remote-registry` is just saying to validate the commit on the inside?

I'm still a bit naive on some of this stuff.

What do you think we should do here for this PR?

Reply:
> What do you think we should do here for this PR?

See https://github.com/ostreedev/ostree-rs-ext/pull/464 which had a similar motivation; instead of forcing everyone to understand the specific-to-ostree-container reference scheme, we split up the options.

(Incidentally note, this raises the point that we've got kickstart syntax in one place, and JSON schema in another...)

Reply:
Made a change to use `ostree-remote-registry` as the default, but signature verification can be disabled by setting the `target_container_signature_verification` option to false. @cgwalters any thoughts?

Reply:
I also think we need to add a check within the deploy stage code to ensure that the `remote` is not empty when we're using `ostree-remote-registry` and that `remote` is not specified when we're using `ostree-unverified-registry`. The optimal place to for this check would be in the deploy stage schema, but it doesn't really work out since we still need to consider the use of `remote` for the ostree commit input case.

Reply:
I'm not sure. Basically there are two levels of "checking" that could be done and I think we need to support all combinations. If someone specifies an ostree remote I think we should just pass that through.

Reply:
> Basically there are two levels of "checking" that could be done and I think we need to support all combinations. 

This is true, but the image reference scheme only supports one level of things.  If you specify an ostree remote, we're not going to also care about the containers-level signature stuff.  If you don't specify an ostree remote, it's only the containers level that matters.

Or to reverse things, there's no explicit way to say "verify with an ostree remote but not containers" or "verify with an ostree remote *and* containers".

Reply:
Shifting away from a boolean might work to catch it in the schema using `implication`. For example, adding this to the schema might work...
```
"anyOf": [
    {
      "not": {
        "properties": { "target_signature_verification": { "const": "ostree-remote" } },
        "required": ["target_signature_verification"]
      }
    },
    { "required": ["remote"] }
  ]
```

https://json-schema.org/understanding-json-schema/reference/conditionals#implication

Reply:
Never mind. Even with my suggested change, it's still the same problem of enforcing `remote` in the schema. We can't default to `target_signature_verification` to be `ostree-remote` when it isn't defined and require `remote`. This is because if we take in an ostree commit and don't define `target_signature_verification`, the code should pass even if remote isn't specified. Therefore, we might need to just settle with it bailing out at runtime.

However, @cgwalters do you have any thoughts on moving away from a boolean?
> `target_signature_verification`: `ostree-remote` (default) - requires remote to be specified.
> `target_signature_verification`: `none` (disabled completely)
> `target_signature_verification`: `container-signing`

Reply:
Sure, that tristate seems fine to me.

Reply:
When I run `rpm-ostree upgrade` when we deploy using `ostree-image-signed:registry:{image_name}`, I get the following error:
```
$ rpm-ostree status
State: idle
Deployments:
● ostree-image-signed:docker://quay.io/fedora/fedora-coreos:stable
               	Digest: sha256:d6ae5e4566d4ddd937e135eb3400db9c78aa76310e7309e601fa8f570d817519
              	Version: 38.20231002.3.1 (2023-10-17T02:25:40Z)

$ sudo rpm-ostree upgrade
Pulling manifest: ostree-image-signed:docker://quay.io/fedora/fedora-coreos:stable
error: Preparing import: Fetching manifest: containers-policy.json specifies a default of `insecureAcceptAnything`; refusing usage
```
Sorry, not 100% sure I understand container signing but I assume it's failing because we need to configure/enable some signing mechanism within `containers-policy.json` locally on the system?

Reply:
> Sorry, not 100% sure I understand container signing but I assume it's failing because we need to configure/enable some signing mechanism within `containers-policy.json` locally on the system?

Yes, two things:

1. what you mentioned above.. The client isn't configured with enough info for `quay.io/fedora/fedora-coreos`

Reply:
> but we're reusing the existing skopeo input/source for this so it would need to make sense in a container context too.

Not necessarily.

osbuild inputs can have additional "metadata" specified, which are stage-specific. The `org.osbuild.files` input has it in its schema https://github.com/osbuild/osbuild/blob/b32ddc41367b0a2b6c771b276dde441e28d88726/inputs/org.osbuild.files#L25-L34 and we use it in the `org.osbuild.rpm` stage to specify whether to check the GPG signature for each input (RPM)  https://github.com/osbuild/osbuild/blob/b32ddc41367b0a2b6c771b276dde441e28d88726/stages/org.osbuild.rpm#L305

example use in a manifest:
https://github.com/osbuild/manifest-db/blob/4076e4939970ca8642fa58c0d6e94d60d74c698c/manifest-db/centos_8-aarch64-ami-boot.json#L22-L29

So the `containers` input could be extended to also optionally contain metadata specific to this stage.